cpp/src/strings/translate.cu

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/translate.hpp>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+#include <algorithm>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+using translate_table = thrust::pair<char_utf8,char_utf8>;
+
+namespace
+{
+
+/**
+ * @brief This is the translate functor for replacing individual characters
+ * in each string.
+ */
+struct translate_fn
+{
+    column_device_view const d_strings;
+    translate_table const* table;
+    size_type table_size;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    __device__ size_type operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        size_type bytes = d_str.size_bytes();
+        char* out_ptr = nullptr;
+        if( d_offsets )
+            out_ptr = d_chars + d_offsets[idx];
+        for( auto itr = d_str.begin(); itr != d_str.end(); ++itr )
+        {
+            auto chr = *itr;
+            auto entry = thrust::find_if( thrust::seq, table, table+table_size,
+                [chr] __device__ ( auto te ) { return te.first==chr; } );
+            if( entry < table + table_size )
+            {
+                bytes -= bytes_in_char_utf8(chr);
+                chr = entry->second;
+                if( chr ) // if null, remove character
+                    bytes += bytes_in_char_utf8(chr);
```

Comment:
This seems incredibly confusing - at first it looks like we are undo-ing something we just did. However, once you realize that we are reassigning to `chr`  - we realize that isn't so. So why are we reassigning? Why not just use `entry->second` or assign to a new variable?

Reply:
This is to make line 73 work: https://github.com/rapidsai/cudf/blob/642a6c8d84c40497331a937265363ef45a74f3ff/cpp/src/strings/translate.cu#L73
The character is either replace or not. If not, the character falls through to the output. If so, the new character is placed in the output.